### ISSUE TITLE
NetworkPolicy based on PodSelector and NamespaceSelector doesn't work.

### ISSUE THREAD

**Issue by TomatoAres** [projectcalico/calico#2993]
NetworkPolicy based on PodSelector and NamespaceSelector doesn't work.
## Expected Behavior

* policy:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  namespace: network-policy
  creationTimestamp: null
  name: allow-client-b-in-ns-b-via-ns-selector-and-pod-selector
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          ns-name: network-policy-c
      podSelector:
        matchLabels:
          pod-name: client-b
  podSelector:
    matchLabels:
      pod-name: network-policy-server
```
* namespace:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2019-11-15T05:52:09Z"
  labels:
    ns-name: network-policy-b
  name: network-policy-b
```
* pod
```
apiVersion: v1
kind: Pod
metadata:
  namespace: network-policy-b
  generateName: client-a-

  labels:
    pod-name: client-b
spec:
  containers:
  - args:
    - /bin/sh
    - -c
    - for i in $(seq 1 5); do nc -vz -w 8 svc-network-policy-server.network-policy
      80 && exit 0 || sleep 1; done; exit 1
    image: beyond.io:5000/busybox:1.29
    name: client-a-container
    resources: {}
  restartPolicy: Never
status: {}

```
this pod was expected to connect to server，but can not.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Calico version : v3.1.0
* Orchestrator version (e.g. kubernetes, mesos, rkt):kubernetes-v1.15.1
* Operating System and version: Centos 7 1804
* Link to your project (optional):

**Comment by tmjd** [comment]
You listed Calico version v3.1.0, was that a typo? If not please try again with the latest 3.10 release as support for namespace selectors was added in v3.10.0.

**Comment by TomatoAres** [comment]
@tmjd Thank you for your answer. I will have a try,if it works ,I will closed this issue.

**Comment by xieyanker** [comment]
> You listed Calico version v3.1.0, was that a typo? If not please try again with the latest 3.10 release as support for namespace selectors was added in v3.10.0.

@tmjd Hi! May I ask if it is supported in version 3.10.0 and above. I meet same issue in calico 3.8.0. I thought calico had already supported it. 
This is my issue: https://github.com/kubernetes/kubernetes/issues/85539

**Comment by TomatoAres** [comment]
@xieyanker I have already check it,v3.10.0 works. [calico v3.10 release note](https://docs.projectcalico.org/v3.10/release-notes/)

**Comment by xieyanker** [comment]
@TomatoAres Thanks, I will upgrade calico to 3.10.1

**Comment by xieyanker** [comment]
@TomatoAres @tmjd @fasaxc 
I have upgraded calico to v3.10.1. However, not works for me still!
```shell
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl get ns -A -owide --show-labels|grep test
test1             Active   11d   ns=test1
test2             Active   11d   <none>
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl get pod -A -owide --show-labels|grep test-nginx
test1             test-nginx1                                         1/1     Running             0          10m     10.151.162.39     master1   <none>           <none>            app=test-nginx1
test2             test-nginx2                                         1/1     Running             0          10m     10.151.208.79     master2   <none>           <none>            app=test-nginx2
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl get networkpolicy --all-namespaces -oyaml
apiVersion: v1
items:
- apiVersion: extensions/v1beta1
  kind: NetworkPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"extensions/v1beta1","kind":"NetworkPolicy","metadata":{"annotations":{},"name":"test-nginx-policy","namespace":"test2"},"spec":{"ingress":[{"from":[{"namespaceSelector":{"matchLabels":{"ns":"test1"}},"podSelector":{"matchLabels":{"app":"test-nginx1"}}}]}],"podSelector":{"matchLabels":{"app":"test-nginx2"}}}}
    creationTimestamp: "2019-12-03T12:32:07Z"
    generation: 1
    name: test-nginx-policy
    namespace: test2
    resourceVersion: "8690440"
    selfLink: /apis/extensions/v1beta1/namespaces/test2/networkpolicies/test-nginx-policy
    uid: eb68fa16-15c8-11ea-b1d2-fa163e59946d
  spec:
    ingress:
    - from:
      - namespaceSelector:
          matchLabels:
            ns: test1
        podSelector:
          matchLabels:
            app: test-nginx1
    podSelector:
      matchLabels:
        app: test-nginx2
    policyTypes:
    - Ingress
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
root@inspurtest14-wxhqn85ukg-master-1:~# calicoctl get networkpolicy --all-namespaces -oyaml
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
apiVersion: projectcalico.org/v3
items:
- apiVersion: projectcalico.org/v3
  kind: NetworkPolicy
  metadata:
    creationTimestamp: 2019-12-03T12:32:07Z
    name: knp.default.test-nginx-policy
    namespace: test2
    resourceVersion: "8690442"
    uid: 5506ca0a-a85f-483e-912e-b0e20b8d215e
  spec:
    ingress:
    - action: Allow
      destination: {}
      source:
        namespaceSelector: ns == 'test1'
        selector: projectcalico.org/orchestrator == 'k8s' && app == 'test-nginx1'
    order: 1000
    selector: projectcalico.org/orchestrator == 'k8s' && app == 'test-nginx2'
    types:
    - Ingress
kind: NetworkPolicyList
metadata:
  resourceVersion: "8692292"
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl exec -it test-nginx1 sh -ntest1
# ping 10.151.208.79
PING 10.151.208.79 (10.151.208.79) 56(84) bytes of data.
^C
--- 10.151.208.79 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 22ms

# 
root@inspurtest14-wxhqn85ukg-master-1:~# calicoctl version
Client Version:    v3.10.1-2-gb82f07ab
Git commit:        b82f07ab
Cluster Version:   v3.10.1-5-g2826afa
Cluster Type:      kubespray,bgp,k8s

```

**Comment by xieyanker** [comment]
Is there any configuration that needs attention？ Maybe I use some wrong configs.

**Comment by tmjd** [comment]
Did you ensure that your test pods have the appropriate `app` labels?
Have you verified without the policy (or any other) that the traffic you are attempting does work?

**Comment by xieyanker** [comment]
@tmjd  Yeah, my pods have the appropriate app labels
```shell
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl get ns -A -owide --show-labels|grep test
test1             Active   11d   ns=test1
test2             Active   11d   <none>
root@inspurtest14-wxhqn85ukg-master-1:~# kubectl get pod -A -owide --show-labels|grep test-nginx
test1             test-nginx1                                         1/1     Running             0          10m     10.151.162.39     master1   <none>           <none>            app=test-nginx1
test2             test-nginx2                                         1/1     Running             0          10m     10.151.208.79     master2   <none>           <none>            app=test-nginx2
```
And if I delete the networkpolicy, I can connect success.

**Comment by tmjd** [comment]
Sorry I should have scrolled over in your first post and seen the labels.
Actually I realized that I was incorrect in suggesting the upgrade, the upgrade would only be required if you were using a Calico GlobalNetworkPolicy and you were using namespaceSelector as the 'top' level selector like here https://docs.projectcalico.org/v3.10/security/namespace-policy#use-kubernetes-rbac-to-control-namespace-label-assignment.

I'm not sure what to say, your setup looks correct. Would you please open another issue since the original issue here was resolved. Please include info about your setup and reference this issue.
I think we might need you to set debug for node/felix and collect logs from the host where nginx2 is running.

**Comment by xieyanker** [comment]
@tmjd Thanks for your reply! I have resolved it! 
My `cni-plugin`‘s configuration didn't start `policy`, it works when I add follows
```shell
      "policy": {
        "type": "k8s"
      },
```
on `/etc/cni/net.d/calico.conflist.template`, refer to [cni-plugin configuration](https://docs.projectcalico.org/v3.10/reference/cni-plugin/configuration).

By the way, I got some error about `authority` when I run new pod after I update this config. It works util I exec `kubectl edit clusterrole system:node` and add `namespaces`'s  `authority`. Is there exists any good idea to resolve it?

![error](https://user-images.githubusercontent.com/45066503/70236550-199ac100-17a0-11ea-91cf-3c433732dea3.png)
